### ISSUE TITLE
Inductor all reduce lowering doesnt require contiguity

### ISSUE THREAD

**Issue by fegin** [pytorch/pytorch#137259]
### 🐛 Describe the bug

`all_reduce()` requires the buffer to be contiguous. However, we have observed that after applying `CompiledAutograd()`, the `all_reduce()` may fail because the input buffer becomes non-contiguous. Below is a minimal repository. You can reproduce the error using the command `torchrun --nproc-per-node 2 test.py`. There are two issues in this example:

1. If `CompiledAutograd()` is removed, there is no error, even without the `contiguous()` call.
2. If `CompiledAutograd()` is added, `all_reduce()` reports an error due to the non-contiguous buffer, even with the `contiguous()` call.

```
import os
import functools
import torch
import torch.nn as nn
import torch.distributed as dist
from torch._dynamo import compiled_autograd
from torch.distributed.distributed_c10d import _get_default_group

FILE_PATH = "/tmp/chienchin_rendezvous"
DIM=2000

class Net(nn.Module):
    def __init__(self, checkpoint=False):
        super().__init__()
        self.fc1 = nn.Linear(DIM, DIM)

    def forward(self, x):
        return self.fc1(x)

def main():
    torch.manual_seed(10)
    local_rank = int(os.environ["LOCAL_RANK"])
    rank = int(os.environ["RANK"])
    world_size = int(os.environ["WORLD_SIZE"])

    torch.cuda.set_device(f"cuda:{local_rank}")
    device = torch.device("cuda")

    import torch.distributed._functional_collectives as fcol
    def hook(param, pg):
        param.grad = fcol.all_reduce(param.grad.contiguous(), "avg", pg)

    dist.init_process_group(
        backend="nccl",
        rank=rank,
        world_size=world_size,
        store=dist.FileStore(FILE_PATH, world_size),
    )
    model = Net().to(device)
    model.fc1.weight.register_post_accumulate_grad_hook(
        functools.partial(hook, pg=_get_default_group())
    )
    batch = torch.randn([1, DIM], device=device)
    with torch._dynamo.utils.maybe_enable_compiled_autograd(True):
        for i in range(2):
            loss = model(batch).sum()
            loss.backward()

if __name__ == "__main__":
    main()
 ```

### Versions

Nightly

cc @ezyang @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @xmfan

**Comment by xmfan** [comment]
Corresponding inductor code:
```
def call(args):
    arg0_1, = args
    args.clear()
    assert_size_stride(arg0_1, (2000, 2000), (2016, 1)) 
    ...
```
Why is args0_1 strided?

**Comment by xmfan** [comment]
oh i think _empty_strided_cuda never marks the tensors as contiguous -.-

**Comment by xmfan** [comment]
`tensor.is_contiguous` doesn't seem to check strides
```
>>> empty_strided_cuda((2000, 2000), (2016, 1), torch.float32).stride()                                                                                     
(2016, 1)                                                                                                                                                   
>>> empty_strided_cuda((2000, 2000), (2016, 1), torch.float32).is_contiguous()                                                                              
False 
```

**Comment by eellison** [comment]
>>> empty_strided_cuda((2000, 2000), (2000, 1), torch.float32).is_contiguous()

It's not contiguous because the last dim strides are not equal to the product of the all the sizes before it. 

Something about the all reduce is modifying the strides of the matmul input. I'm not sure why. If you return directly after `mm = torch.ops.aten.mm.default(permute, arg1_1);` in [repro](https://gist.github.com/eellison/48a5f94b7099e45dc44a63ca513f1a37) we will return a contiguous output.

all_reduce could potentially use the layout constraint `require_contiguous`, or there might be something else going on.

**Comment by eellison** [comment]
Fix here https://github.com/pytorch/pytorch/pull/137345
